Torus mesh generator

anianruoss opened the pull request:
I implemented the TorusMeshBuilder to the best of my abilities (see images below). Please check if everything looks correct @hiptmair.

Two issues arose:

- I declared/defined the class in the file `tp_triag_mesh_builder.h`/`tp_triag_mesh_builder.h` because all other the builders are also there. However, it would be nicer to have a separate file for every builder. What do you think @craffael?
- I didn't document the example with Doxygen yet, as I am waiting for your response @craffael.


**R = 4 * r, NoXCells = NoYCells = 4**
![torus_4](https://user-images.githubusercontent.com/18675644/47260417-86c79600-d4bb-11e8-9486-bd0f3eca6952.png)

**R = 4 * r, NoXCells = NoYCells = 4 after 4 uniform refinement steps**
![torus_4_refined](https://user-images.githubusercontent.com/18675644/47260448-07869200-d4bc-11e8-8518-39291f043228.png)

**R = 4 * r, NoXCells = NoYCells = 50**
![torus_50](https://user-images.githubusercontent.com/18675644/47260451-10776380-d4bc-11e8-8d87-02bc9812efcc.png)

**R = 2 * r, NoXCells = NoYCells = 50**
![figure_1](https://user-images.githubusercontent.com/18675644/47260512-4e28bc00-d4bd-11e8-95fd-342d4e3d22a3.png)

hiptmair commented on lib/lf/mesh/test/tp_triag_mesh_builder_tests.cc:
```diff
@@ -120,4 +120,34 @@ TEST(lf_mesh_p, buildTPQuadMesh) {
+  std::cout << "Printing mesh information" << std::endl;
+  utils::PrintInfo(*mesh_p, std::cout);
+}
+
```
Augment the test by checking that every edge has exactly two adjacent cells and that every vertex has exactly four adjacent edges and cells. To do so, you can use the function `countNoSuperEntities()`.

anianruoss replied:
Done, thank you for the suggestion :+1:

craffael commented:
>     * I declared/defined the class in the file `tp_triag_mesh_builder.h`/`tp_triag_mesh_builder.h` because all other the builders are also there. However, it would be nicer to have a separate file for every builder. What do you think @craffael?
> 

I agree with you. I think we should follow the one-file-per-class rule which is encouraged by the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#File_Names). The big advantage of doing so is that it is always clear where to look for a particular class. With the current situation, if someone from outside is looking for the TorusMeshBuilder he will probably not find it straight away.

Therefore, I suggest that we split the files `tp_triag_mesh_builder.h` and `tp_triag_mesh_builder.cc` into one header/source file per class. Moreover I suggest to move these files into the `lf::mesh::utils` module. What do you think @hiptmair ? @anianruoss Could you do the refactoring?

>     * I didn't document the example with Doxygen yet, as I am waiting for your response @craffael.

I will comment on https://github.com/craffael/lehrfempp/pull/52#issuecomment-427921913 in a moment.

Finally I want to point out, that the VtkWriter should be able to handle the torus mesh and I would really encourage you to try it. Paraview is much better when it comes to 3D meshes. If the VtkWriter doesn't work, let me know and I will fix it.

craffael commented on lib/lf/mesh/test/tp_triag_mesh_builder_tests.cc:
```diff
@@ -120,4 +120,34 @@ TEST(lf_mesh_p, buildTPQuadMesh) {
+  EXPECT_EQ(mesh_p->Size(1), 30) << "Mesh should comprise 30 edges";
+  EXPECT_EQ(mesh_p->Size(2), 15) << "Mesh should have 15 vertices";
+
+  std::cout << "Checking entity indexing" << std::endl;
```
Personally I'm not a big fan of writing to std::cout in tests because it convolutes the output considerably. Check e.g. the output that is produced by travis-ci for every commit... I think a test should only write to the command line if something goes wrong, which is exactly what you do on lines 138-143.

I know that at the moment the code is not very consistent on this, but this is something  would like to fix eventually.

anianruoss replied:
Agreed :+1:

anianruoss replied:
Agreed :+1:

anianruoss commented on lib/lf/mesh/tp_triag_mesh_builder.cc:
```diff
@@ -243,4 +244,106 @@ std::shared_ptr<mesh::Mesh> TPQuadMeshBuilder::Build() {
+      node_coord << (R + r * std::cos(theta(i))) * std::cos(phi(j)),
+          (R + r * std::cos(theta(i))) * std::sin(phi(j)),
+          r * std::sin(theta(i));
+
```
Great, thanks :blush:

anianruoss commented on lib/lf/mesh/tp_triag_mesh_builder.cc:
```diff
@@ -243,4 +244,106 @@ std::shared_ptr<mesh::Mesh> TPQuadMeshBuilder::Build() {
+      t_idx[quad_cnt] = mesh_factory_->AddEntity(
+          lf::base::RefEl::kQuad(), vertex_index_list,
+          std::make_unique<geometry::QuadO1>(quad_geo));
+    }
```
@hiptmair I'm sorry, but I'm not quite sure if this is just a comment or a change request?

hiptmair replied:
This was not intended as a change request. The current implementation is fine.

anianruoss commented:
Thank you both for having a look. I will implement the changes you requested.

> Therefore, I suggest that we split the files tp_triag_mesh_builder.h and tp_triag_mesh_builder.cc into one header/source file per class. Moreover I suggest to move these files into the lf::mesh::utils module. What do you think @hiptmair ? @anianruoss Could you do the refactoring?

If @hiptmair agrees I'm happy to do the refactoring :+1:

hiptmair commented:
It is certainly a good idea to move the implementation for the torus into separate files. 
As I have expressed in [Issue #57](https://github.com/craffael/lehrfempp/issues/57), I am not so fond of the many subdirectories of `mesh`, but this can be fixed later.

anianruoss commented:
I implemented the requested changes. The VTK output looks really cool :blush:

Once again two issues:
- for some reason, the Travis and AppVeyor integrations stopped testing my commits
- I moved the tests into the directory `lf/mesh/utils/test` instead of `lf/mesh/test_utils`. In my opinion it would be cleaner to have all the tests from `lf/mesh/test_utils` in the `lf/mesh/utils/test` directory. Should I change this?

@craffael @hiptmair ptal

![torus_3](https://user-images.githubusercontent.com/18675644/47304815-f8245780-d627-11e8-8bfc-09450c58fd6e.png)
![torus_50_flat](https://user-images.githubusercontent.com/18675644/47304816-f8245780-d627-11e8-866f-30c0dc7b3a26.png)
![torus_50_upright](https://user-images.githubusercontent.com/18675644/47304817-f8245780-d627-11e8-89cb-7580288a4684.png)

craffael commented:
Can you merge the pull request?

anianruoss commented:
@craffael I was waiting for your approval of the new changes

anianruoss commented:
@craffael I merged the pull request below and had to change the branch.

craffael commented:
I'm sorry, I've setup something wrong. You should now be able to merge it...

anianruoss commented:
@craffael what about this:

> I moved the tests into the directory lf/mesh/utils/test instead of lf/mesh/test_utils. In my opinion it would be cleaner to have all the tests from lf/mesh/test_utils in the lf/mesh/utils/test directory. Should I change this?

craffael commented:
Yes thats fine, the tests for the torus mesh generator really belong to lf/mesh/utils/test`. The folder `lf/mesh/test_utils` is a "test utility library" that can be included by other test targets to reuse testing functionality. (see also https://github.com/craffael/lehrfempp/wiki/Contribute#221-test-modules)
